hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/EnrichmentProcessor.java

```diff
@@ -43,13 +56,22 @@ public void process(StructuredTrace trace) {
         trace, enrichmentArrivalTimer, ENRICHMENT_ARRIVAL_TIME);
     AvroToJsonLogger.log(LOG, "Structured Trace before all the enrichment is: {}", trace);
     for (Enricher enricher : enrichers) {
+      FutureTask<Void> task =
+          new FutureTask<>(
+              () -> {
+                applyEnricher(enricher, trace);
+                return null;
+              });
       try {
-        applyEnricher(enricher, trace);
+        executorService.submit(task);
+        task.get(enricherTaskTimeout.toSeconds(), TimeUnit.SECONDS);
       } catch (Exception e) {
+        boolean cancelStatus = task.cancel(true);
         LOG.error(
-            "Could not apply the enricher: {} to the trace with traceId: {}",
+            "Could not apply the enricher: {} to the trace with traceId: {}. Task is cancelled with status {}",
```

Comment:
This error applies to both cancelled and throwing enrichers, right? That doesn't seem clear to me from the message since the task may or may not have been cancelled, perhaps a separate catch?

Reply:
Along with streamlining the task itself, maybe something like:
```diff
diff --git a/hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/EnrichmentProcessor.java b/hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/EnrichmentProcessor.java
index f632932..014b978 100644
--- a/hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/EnrichmentProcessor.java
+++ b/hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/EnrichmentProcessor.java
@@ -8,7 +8,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.concurrent.FutureTask;
+import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import org.hypertrace.core.datamodel.Edge;
 import org.hypertrace.core.datamodel.Entity;
@@ -56,22 +56,22 @@ public class EnrichmentProcessor {
         trace, enrichmentArrivalTimer, ENRICHMENT_ARRIVAL_TIME);
     AvroToJsonLogger.log(LOG, "Structured Trace before all the enrichment is: {}", trace);
     for (Enricher enricher : enrichers) {
-      FutureTask<Void> task =
-          new FutureTask<>(
-              () -> {
-                applyEnricher(enricher, trace);
-                return null;
-              });
+      Future<?> task = executorService.submit(() -> applyEnricher(enricher, trace));
       try {
-        executorService.submit(task);
         task.get(enricherTaskTimeout.toSeconds(), TimeUnit.SECONDS);
-      } catch (Exception e) {
+      } catch (InterruptedException interruptedException) {
         boolean cancelStatus = task.cancel(true);
         LOG.error(
-            "Could not apply the enricher: {} to the trace with traceId: {}. Task is cancelled with status {}",
+            "Enricher: {} timed out after {}s for the trace with traceId: {}. Task is cancelled with status {}",
+            enricher.getClass().getCanonicalName(),
+            enricherTaskTimeout.toSeconds(),
+            HexUtils.getHex(trace.getTraceId()),
+            cancelStatus);
+      } catch (Exception e) {
+        LOG.error(
+            "Could not apply the enricher: {} to the trace with traceId: {}",
             enricher.getClass().getCanonicalName(),
             HexUtils.getHex(trace.getTraceId()),
-            cancelStatus,
             e);
       }
     }
```

Reply:
Separated them. However I kept task submission in try block as it can throw `RejectedExecutionException`